Add new SQLCheckOperators

denimalpaca opened the pull request:
This pull request adds two new SQL Check Operators to the core sql.py file, one called SQLColumnCheckOperator and one called SQLTableCheckOperator. Corresponding tests are added as well. The current functionality of these operators is minimal, but expected to grow over time. These operators, unlike current ones, save information about the results
of the check queries that is then parseable by third party libraries.

closes: #23874

Commit:
Add new SQLCheckOperators

This commit adds two new SQL Check Operators to the core sql.py file,
one called SQLColumnCheckOperator and one called SQLTableCheckOperator.
Corresponding tests are added as well. The current functionality of
these operators is minimal, but expected to grow over time. These
operators, unlike current ones, save information about the results
of the check queries that is then parseable by third party libraries.

uranusjr commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,218 @@ def push(self, meta_data):
+                f"Pass Value: {check_values['pass_value']}, "
+                f"Result: {check_values['result']}\n"
+            )
+    return failed_tests
```
```suggestion
def _get_failed_tests(checks):
    return [
        f"\tCheck: {check}, "
        f"Pass Value: {check_values['pass_value']}, "
        f"Result: {check_values['result']}\n"
        for check, check_values in checks.items()
        if not check_values["success"]
    ]
```

uranusjr replied:
But this string does not seem to be used at all. Maybe it should simply be like this?

```python
def _all_success(checks):
    return all(values["success"] for values in checks.values())
```

denimalpaca replied:
The string is used in line 575 in the `AirflowException`. Great suggestion using the list comprehension, thank you!

Commit:
Use list comprehension in _get_failed_tests()

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Commit:
Remove repeated lines of code

Accidentally copied the last line of the file several times, this
commit remedies that change.

Commit:
Fix doc strings.

Commit:
Update SQLColumnCheckOperator

Improved the column check operator to include support for
greater than, geq, less than, and leq functions.
Updated logic and tests to reflect this change.

Commit:
Update SQLTableCheckOperator

Improved the table check operator by refactoring
how checks are performed; this makes it much more customizeable
at the expense of knowing the specific result of checks, as
the results are now all just 0 or 1.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+                'tolerance': 0.01
+            }
+        }
+    }
```
For these pesky build docs errors, you can try adding a `.. code-block::` directive and house these example dicts under them. Something like:
```rst
:param table: the table to run checks on.
:param column_mapping: the dictionary of columns and their associated checks, e.g.:

    .. code-block::

        column_mapping = {
            'col_name': {
                'null_check': {
                    'pass_value': 0,
                },
                'min': {
                    'pass_value': 5,
                    'tolerance': 0.2,
                }
            }
        }

:param conn_id: the connection ID used to connect to the database.
:param database: name of database which overwrite the defined one in connection
```
Check out the docstrings for [`chain()`](https://github.com/apache/airflow/blob/fd4e344880b505cfe53a97a6373d329515bbc7a3/airflow/models/baseoperator.py#L1520) and [`cross_downstream()`](https://github.com/apache/airflow/blob/fd4e344880b505cfe53a97a6373d329515bbc7a3/airflow/models/baseoperator.py#L1649) for examples.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+
+    column_checks = {
+        # pass value should be number of acceptable nulls
+        "null_check": "SUM(CASE WHEN 'column' IS NULL THEN 1 ELSE 0 END) AS column_null_check",
```
```suggestion
        "null_check": "SUM(CASE WHEN column IS NULL THEN 1 ELSE 0 END) AS column_null_check",
```
Should be checking for the column value rather than the string of the the column name I think?

denimalpaca replied:
Ummm... good question. I'm not 100% sure. Testing on Snowflake, the lack of `'` version had a parse error iirc. So it might be a dialect issue that can be resolved in the provider versions.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+                    self.column_mapping[column][checks[idx]]["tolerance"]
+                    if "tolerance" in self.column_mapping[column][checks[idx]]
+                    else None
+                )
```
```suggestion
                tolerance = self.column_mapping[column][checks[idx]].get("tolerance")
```

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+            and "equal_to" not in check_values
+        ):
+            raise ValueError(
+                "Please provide one or more of: less_than, leq_than, "
+                "greater_than, geq_than, or equal_to in the check's dict."
```
I think it would be super helpful if listing these equality check options were in the docstring so they can appear in the Python API documentation for users.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+
+        self.table = table
+        self.column_mapping = column_mapping
+        self.sql = f"SELECT * FROM {self.table};"
```
Just curious, why is this needed here?

denimalpaca replied:
For OpenLineage. The Extractor needs some static query to render, unless we want to have the entire templated query rendered in `__init__()`. I can double-check this, though, it might have changed with the move to the Listener.

potiuk replied:
Just to reiterate - those "common" SQL classes will not be released unti Airflow 2.4 and it means that the first operator tha will be able to use them as "mandatory" will be 12 months after 2.4 is released (that's our policy for providers), so I just want to make you aware @denimalpaca that those classes won't be to useful for quite some time. 

Maybe a better Idea will be to add a separate "base sql" provider which will not be in the "airflow/operators" but it will be a separate package that could be installed additionally to airflow. We have not done so before, but maybe there is a good reason to introduce such a "shared sql" provider that wil encapsulate all SQL-like base functionality that open lineage might base on? I think trying to implement it in Airflow core is a bad idea, if the goal is fast adoption by multiple providers. 

Just saying.

josh-fell replied:
>Maybe a better Idea will be to add a separate "base sql" provider

I think this is an interesting idea and I support it. We could start with these operators and slowly move existing core SQL operators over too.

@potiuk Do you think this provider would be included with Airflow installs like HTTP, FTP, SQLite, and IMAP, or completely separate?

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+
+            failed_tests.extend(_get_failed_tests(self.column_mapping[column]))
+        if failed_tests:
+            raise AirflowException(
```
Maybe it's worth thinking about having a separate exception for Airflow checks/test especially if more data quality functionality is coming?

denimalpaca replied:
I think that's a good idea, but we'd also have to replace exceptions in other SQL operators and I think that's outside the scope of this PR. But I can open an issue and implement that in a different PR. Something like an `AirflowDataQualityException`?

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+        if "greater_than" in check_values and "less_than" in check_values:
+            if check_values["greater_than"] >= check_values["less_than"]:
+                raise ValueError(
+                    "greater_than should be strictly less than to "
```
```suggestion
                    "greater_than should be strictly less than "
```

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+
+        self.table = table
+        self.checks = checks
+        self.sql = f"SELECT * FROM {self.table};"
```
Same question here as above.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+            self.log.info(f"Record: {records}")
+
+            if not records:
+                raise AirflowException("The query returned None")
```
```suggestion
            if not records:
                raise AirflowException("The query returned None")

            self.log.info(f"Record: {records}")
```
Seems like it's not worth logging the record info if no records are returned and a failure will occur. WDYT about flipping these around?

Also, WDYT about changing the exception message to something like?
"The following query returned no rows: {sql}"

denimalpaca replied:
Ah, good points. Yeah, I think that would be a better order and message.

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+            checks = [*self.column_mapping[column]]
+            checks_sql = ",".join([self.column_checks[check].replace("column", column) for check in checks])
+
+            self.sql = f"SELECT {checks_sql} FROM {self.table};"
```
Similar to the question above, does this need to be an instance attr?

josh-fell commented on airflow/operators/sql.py:
```diff
@@ -467,6 +467,269 @@ def push(self, meta_data):
+        self.log.info(f"Record: {records}")
+
+        if not records:
+            raise AirflowException("The query returned None")
```
Same comment as above re: flipping these around.

josh-fell commented on tests/operators/test_sql.py:
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
             operator.execute()
 
 
+class TestColumnCheckOperator(unittest.TestCase):
```
IMO this can be a `pytest` test rather than continuing to use `unittest`.

denimalpaca replied:
Is there a reason to use one over the other? I was just copying the test pattern from the other ones.

potiuk replied:
We prefer pytest test type unless there are good reasons not to (for new test)

josh-fell commented on tests/operators/test_sql.py:
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
+    def _construct_operator(self, column_mapping):
+        return SQLColumnCheckOperator(task_id="test_task", table="test_table", column_mapping=column_mapping)
+
+    @mock.patch.object(SQLColumnCheckOperator, "get_db_hook")
```
Since every test case uses the same mocked object, you could think about using a pytest fixture for mocking `get_db_hook`. I don't have a strong opinion though.

uranusjr replied:
If you use pytest fixtures, the `monkeypatch` fixture makes this easy to do in `setup_method`.

josh-fell commented on tests/operators/test_sql.py:
```diff
@@ -385,6 +387,82 @@ def test_fail_min_sql_max_value(self, mock_get_db_hook):
+        with pytest.raises(AirflowException, match="Invalid column check: invalid_check_name."):
+            self._construct_operator(self.invalid_column_mapping)
+
+    @mock.patch.object(SQLColumnCheckOperator, "get_db_hook")
+    def test_pass_all_checks_exact_check(self, mock_get_db_hook):
+        mock_hook = mock.Mock()
+        mock_hook.get_first.return_value = (0, 10, 10, 1, 19)
+        mock_get_db_hook.return_value = mock_hook
+        operator = self._construct_operator(self.valid_column_mapping)
+        operator.execute()
+
+    @mock.patch.object(SQLColumnCheckOperator, "get_db_hook")
+    def test_pass_all_checks_inexact_check(self, mock_get_db_hook):
+        mock_hook = mock.Mock()
+        mock_hook.get_first.return_value = (0, 9, 12, 0, 15)
+        mock_get_db_hook.return_value = mock_hook
+        operator = self._construct_operator(self.valid_column_mapping)
+        operator.execute()
```
What is the assertion being made in these happy-path tests?

denimalpaca replied:
It's implicitly `assert mock_hook.get_first.return_value[i] == self.valid_column_mapping[condition_i]`

Commit:
Fix doc errors in docstrings

New docstrings were using example dictionaries in the param list,
and the way they were written before caused doc build errors. The
fix was to update the formatting and add missing words to the word list.

Commit:
Reorder logging

Moves record logging below the AirflowException for no records,
and makes the Exception message more clear.

Commit:
Apply suggestions from code review

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

Commit:
Add more information to docstring

Commit:
Fix docstring indent error

Commit:
Remove reminder comments; Add column alias for unique_check query.

Commit:
Fix SQL rendering bug in SQLTableCheckOperator

SQLTableCheckOperator was rendering an invalid SQL query,
this fix changes the way the query template is built from the
passed in strings.

Commit:
Add note about OpenLineage compatability; add tab in _get_failed_test result string

Commit:
Remove quotes around column in null_check

After some research, it seems the null_check query does *not* need
quotes around the column name.

Commit:
Merge branch 'apache:main' into add_sql_column_table_check_operators
